### PR TITLE
PLT-7540: Automate build process + code-signing for Windows/Mac

### PR DESCRIPTION
**Pull request by MusikPolice**
Before submitting, please confirm you've
 - [x] read and understood our [Contributing Guidelines](https://github.com/mattermost/desktop/blob/master/CONTRIBUTING.md)
 - [x] completed [Mattermost Contributor Agreement](http://www.mattermost.org/mattermost-contributor-agreement/)

**Summary**
A Jenkins Pipeline that is capable of building and code signing the desktop application for Windows, MacOS, and Linux. The corresponding Jenkins job [is here](https://build.mattermost.com/job/mattermost-desktop-pipeline). 

Once merged, the job configuration will be updated so that this runs on any change to any branch in the desktop repository. That way, the existing [mattermost-desktop-cut-release](https://build.mattermost.com/job/md/job/mattermost-desktop-cut-release/configure) can still be used to cut releases. It creates a new branch, which this job will pick up and build.

We may want to think about artifact naming before we start using this to build releases. Currently, this pipeline removes version numbers from all artifacts that it builds, but we might not want to do that, since it could lead to confusion when trying to download the latest artifacts for release testing.

Finally, sorry for all of the commits to this branch. There isn't a great way to test this stuff other than to commit it and run it on Jenkins, since setting up my own Jenkins instance with EC2 spot instances and AMIs and such is a pain. Just look at the final script. The interim commits don't matter, and they'll all be squashed down to one when we merge anyway.

**Issue link**
https://mattermost.atlassian.net/browse/PLT-7540

### CODE REVIEW

**Comment by jasonblais** [comment]
@MusikPolice Is there a rough ETA on when this PR might be ready for review? 

Really excited for this PR :)

**Comment by MusikPolice** [comment]
@jasonblais There are three things left to do:
1. Create a macOS Jenkins slave and move the mac build step to it so that it succeeds (right now, the build always fails because we can't build a macOS release on Jenkins' linux-based slaves)
2. Add support for signing the Windows build. I almost have this working locally. More details in the linked JIRA ticket.
3. Add support for signing the macOS build. This _should_ be as simple as setting some environment variables prior to running the existing build step, but I'm still investigating.

As for timeline, if we just do step 1, this is ready to be deployed, and we can use it to replace Circle CI with Jenkins. Steps 2 and 3 are an enhancement that remove the existing manual code signing step that we have to do with every release.

**Comment by MusikPolice** [comment]
@jasonblais @yuya-oc @ccbrown @AndersonWebStudio this is ready for review.

File: build/Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+                sh 'npm install'
+                sh 'npm run build'
+
+                echo 'Packaging for Linux'
+                sh 'npm run package:linux'
+                sh 'mv release/mattermost-desktop-*-ia32.tar.gz  release/tmp-linux-ia32.tar.gz'
+                sh 'mv release/mattermost-desktop-*.tar.gz release/mattermost-desktop-linux-x64.tar.gz'
+                sh 'mv release/tmp-linux-ia32.tar.gz release/mattermost-desktop-linux-ia32.tar.gz'
+                sh 'mv release/mattermost-desktop_*_i386.deb release/mattermost-desktop-linux-i386.deb'
+                sh 'mv release/mattermost-desktop_*_amd64.deb  release/mattermost-desktop-linux-amd64.deb'
```
Comment (MusikPolice) [inline build/Jenkinsfile]: Like I said in the description of this PR, we might consider removing these `mv` statements and archiving the artifacts with version numbers in their filenames. That could help to keep artifacts sorted during release testing. We'd just have to manually rename them when uploading them to S3 once the release is ready. We could create another Jenkins pipeline to do that too if desired.

**Comment by jasonblais** [comment]
@MusikPolice Is the plan to now only code sign the release branches, similar to what we do for the server?

**Comment by MusikPolice** [comment]
@jasonblais I think so. I'll have to look into how to actually make that secure though. I don't want to just blindly sign any branch that as a "release" prefix on it's name, because any developer can create one. 

We may need to add an optional boolean parameter that controls whether or not the job will sign the builds. This parameter would default to `false`, but when the existing `desktop-cut-release` job is manually triggered, it would kick off this job with the parameter set to `true`. That way, only manually started jobs could actually code sign artifacts.

**Comment by jasonblais** [comment]
@MusikPolice Yeah, that would make sense.

It might also be worthwhile to check how the code signing is done for the server.

**Comment by yuya-oc** [comment]
@MusikPolice Just curious, and not required for now. `electron-builder` has a feature to do code-signing.

https://www.electron.build/code-signing

Are you considering to use it in future? I think that the feature would make scripts simple.

**Comment by jasonblais** [comment]
@AndersonWebStudio I believe you're taking over this PR now. If you have any questions, don't hesitate to let @yuya-oc or Joram know.

File: build/Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    environment {
+        MATTERMOST_DESKTOP_WINDOWS_PFX_PASSWORD = credentials('mattermost-desktop-windows-pfx-password')
+    }
+    stages {
+        stage('Build Linux') {
+            agent any
+            steps {
+                echo 'Building Mattermost Desktop App'
+                sh 'rm -rf release'
+                sh 'npm install'
```
[the remark this answers is not in the document]
Comment (AndersonWebStudio) [reply build/Jenkinsfile]: Updated command to use yarn install, @yuya-oc can you give this another review?

File: build/Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+                  sh '''#!/bin/bash
+                    rm -rf release
+                    yarn install
+                    npm run build
+                  '''
+
+                  echo 'Packaging for MacOS'
+                  sh '''#!/bin/bash
+                    npm run package:mac
+                    MACOS_VERSION=$(awk -F'-' '{print $2 }' <<< $(ls release/Mattermost-*-mac.tar.gz))
```
Comment (yuya-oc) [inline build/Jenkinsfile]: `tar.gz` -> `zip`

File: build/Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+                    echo "MACOS_VERSION is $MACOS_VERSION"
+                    mv release/Mattermost-*-mac.tar.gz release/mattermost-desktop-$MACOS_VERSION-macos.tar.gz
```
Comment (yuya-oc) [inline build/Jenkinsfile]: `tar.gz` -> `zip`

File: build/Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+                  '''
+                  archiveArtifacts artifacts: 'release/mattermost-desktop-*-macos.tar.gz'
```
Comment (yuya-oc) [inline build/Jenkinsfile]: `tar.gz` -> `zip`

File: build/Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,138 @@
+#!/usr/bin/env groovy
```
Comment (yuya-oc) [inline build/Jenkinsfile]: Just curious, is shebang needed? In the Jenkinsfile doc, shebang is not used.

https://jenkins.io/doc/book/pipeline/jenkinsfile/

File: build/Jenkinsfile [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+                '''
+
+                echo 'Packaging for Linux'
+                sh 'npm run package:linux'
+
+                archiveArtifacts artifacts: 'release/mattermost-desktop-*-ia32.tar.gz,release/mattermost-desktop-*[^ia32].tar.gz,release/mattermost-desktop_*_i386.deb,release/mattermost-desktop_*_amd64.deb'
+            }
+        }
+        stage('Build Windows') {
+            agent any
```
Comment (yuya-oc) [inline build/Jenkinsfile]: Just curious, can we use Windows slave in future? We would be able to use Node's native modules.

File: build/Jenkinsfile [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+    environment {
+        MATTERMOST_DESKTOP_WINDOWS_PFX_PASSWORD = credentials('mattermost-desktop-windows-pfx-password')
+    }
+    stages {
+        stage('Build Linux') {
+            agent any
+            steps {
+                echo 'Building Mattermost Desktop App'
+                sh '''#!/bin/bash
```
Comment (yuya-oc) [inline build/Jenkinsfile]: I think that `set -e` is needed for each bash script in order to stop the pipeline as script error when an error occurred in a certain command.

**Comment by yuya-oc** [comment]
@AndersonWebStudio After we merge this, how do we cut new release? Is it enough just pushing a tag?

**Comment by AndersonWebStudio** [comment]
We have moved this pipeline to the shared-pipelines repo, this pull request no longer needs to be merged.

**Pull request closed without merging**